xgi/linalg/matrix.py

```diff
@@ -361,3 +363,149 @@ def clique_motif_matrix(H, index=False):
         return W, rowdict
     else:
         return W
+
+
+def boundary_matrix(S, order=1, orientations=None, index=False):
+    """
+    A function to generate the boundary matrices of an oriented simplicial complex.
+    The rows correspond to the (order-1)-simplices and the columns to the (order)-simplices.
+
+    Parameters
+    ----------
+    S: simplicial complex object
+        The simplicial complex of interest
+    order: int, default: 1
+        Specifies the order of the boundary
+        matrix to compute
+    orientations: dict, default: None
+        Dictionary mapping non-singleton simplices
+        IDs to their boolean orientation
+    index: bool, default: False
+        Specifies whether to output dictionaries
+        mapping the simplices IDs to indices
+
+    Returns
+    -------
+    B: numpy.ndarray
+        The boundary matrix of the chosen order, has dimension
+        (n_simplices of given order - 1, n_simplices of given order)
+    rowdict: dict
+        The dictionary mapping indices to
+        (order-1)-simplices IDs, if index is True
+    coldict: dict
+        The dictionary mapping indices to
+        (order)-simplices IDs, if index is True
+
+    References
+    ----------
+    "Discrete Calculus"
+    by Leo J. Grady and Jonathan R. Polimeni
+    https://doi.org/10.1007/978-1-84996-290-2
+
+    """
+
+    # Extract the simplices involved
+    if order == 1:
+        simplices_d_ids = S.nodes
+    else:
+        simplices_d_ids = S.edges.filterby("order", order - 1)
+
+    if order == 0:
+        simplices_u_ids = S.nodes
+    else:
+        simplices_u_ids = S.edges.filterby("order", order)
+
+    nd = len(simplices_d_ids)
+    nu = len(simplices_u_ids)
+
+    simplices_d_dict = dict(zip(simplices_d_ids, range(nd)))
+    simplices_u_dict = dict(zip(simplices_u_ids, range(nu)))
+
+    if index:
+        rowdict = {v: k for k, v in simplices_d_dict.items()}
+        coldict = {v: k for k, v in simplices_u_dict.items()}
+
+    if orientations == None:
+        orientations = {
+            idd: 0 for idd in list(S.edges.filterby("order", 1, mode="geq"))
+        }
+
+    B = np.zeros((nd, nu))
+    if not (nu == 0 or nd == 0):
+        if order == 1:
+            for u_simplex_id in simplices_u_ids:
+                u_simplex = S.edges.members(u_simplex_id)
+                matrix_id = simplices_u_dict[u_simplex_id]
+                head_idx = list(u_simplex)[1]
+                tail_idx = list(u_simplex)[0]
```

Comment:
`u_simplex` is a set, no? Then I would very much recommend AGAINST converting a set to list and then using the resulting order. How/when is it guaranteed that `list(u_simplex)[1]` is the head and that `list(u_simplex)[0]` is the tail?

Reply:
In order to have orientation, one needs to extract some "reference orientation" for the simplices of the `SimplicialComplex` object. In this implementation, I chose the order in which the nodes appear in the set representing each simplex. If, for some reason, the order of the nodes in the set changes, then it is possible that the orientation of the simplex will be changed too. This can, however, be easily fixed by changing the orientation value in the input dictionary. I see that this is not optimal, but I don't see another simple way to do it without adding a whole new class explicitly designed to account for orientation.

Reply:
The problem is that sets are *not* ordered, there is no such a thing as "the order in which the nodes appear in the set". When you do `list(u_simplex)`, the returned list is not guaranteed to be always the same (at least not across machines/versions/environments). I am actually surprised that this PR passed the automated tests for this reason.

Perhaps we can do `head_idx, tail_idx = max(u_simplex), min(u_simplex)`, or something similar. Here you are taking the reference orientation to be that given by the sorted node IDs. Since most of the time the IDs are integers or strings, this is always well defined.

@maximelucas what do you think?

Reply:
Good point. Now the reference orientation is the one given by the sorted nodes. I think that this should solve the issues if we can assume that node labels are always strings or numbers.